diff --git a/pty_termios.c b/pty_termios.c
index c605b23..ffc0a0d 100644
--- a/pty_termios.c
+++ b/pty_termios.c
@@ -487,7 +487,7 @@ exp_getptymaster()
 	 */
 	if (exp_pty_test_start() == -1) return -1;
 
-#if !defined(HAVE_CONVEX_GETPTY) && !defined(HAVE_PTYM) && !defined(HAVE_SCO_CLIST_PTYS)
+#if !defined(HAVE_CONVEX_GETPTY) && !defined(HAVE_PTYM) && !defined(HAVE_SCO_CLIST_PTYS) && !defined(__MVS__)
 	for (bank = banks;*bank;bank++) {
 		*tty_bank = *bank;
 		*tty_num = '0';
@@ -518,6 +518,22 @@ exp_getptymaster()
         }
 #endif
 
+#if defined(__MVS__)
+        for (num = 0; ; num++) {
+            char num_str [16];
+
+            sprintf (num_str, "%02d", num);
+            sprintf (master_name, "%s%s", "/dev/ptyp00", num_str);
+            if (stat (master_name, &stat_buf) < 0)
+                break;
+            sprintf (slave_name, "%s%s", "/dev/ttyp00", num_str);
+
+            master = exp_pty_test(master_name,slave_name,'0',num_str);
+            if (master >= 0)
+                goto done;
+        }
+#endif
+
 #ifdef HAVE_PTYM
 	/* systems with PTYM follow this idea:
 
